reconcile/utils/oc.py

```diff
@@ -1087,18 +1088,31 @@ def __init__(self, clusters=None, namespaces=None,
             raise KeyError('expected only one of clusters or namespaces.')
         elif clusters:
             threaded.run(self.init_oc_client, clusters, self.thread_pool_size,
-                         cluster_admin=cluster_admin)
+                         privileged=cluster_admin)
         elif namespaces:
-            clusters = []
-            cluster_names = []
+            clusters = {}
+            privileged_clusters = {}
             for ns_info in namespaces:
-                cluster = ns_info['cluster']
-                name = cluster['name']
-                if name not in cluster_names:
-                    cluster_names.append(name)
-                    clusters.append(cluster)
-            threaded.run(self.init_oc_client, clusters, self.thread_pool_size,
-                         cluster_admin=cluster_admin)
+                # init a namespace with clusterAdmin with both auth tokens
+                # OC_Map is used in various places and even when a namespace
+                # declares clusterAdmin token usage, many of those places are
+                # happy with regular dedicated-admin and will request a cluster
+                # with oc_map.get(cluster) without specifying privileged access
+                # specifically
+                c = ns_info['cluster']
+                clusters[c['name']] = c
+                privileged = ns_info.get("clusterAdmin", False) or \
+                    cluster_admin
+                if privileged:
+                    privileged_clusters[c['name']] = c
```

Comment:
Would it make sense to raise an error here if `clusterAdminAutomationToken` is missing from the cluster ? This would allow MR/dry-runs to fail if `clusterAdmin: True` but the token is not configured in the cluster.

Reply:
the integration will fail if a required token is missing (regardless of dedicated-admin or cluster-admin) but it will do whatever work can be done before failing. the path is like this for the openshift-resources integration:

* during oc initialisation, a [missing token results in registering an OCLogMsg](https://github.com/geoberle/qontract-reconcile/blob/c4405e8d90ce45dd371f6f53cc06e8424d84e0af/reconcile/utils/oc.py#L1146-L1161) instead of an oc client
* during state spec initialisation, the problem with a cluster is [detected and registered](https://github.com/geoberle/qontract-reconcile/blob/c4405e8d90ce45dd371f6f53cc06e8424d84e0af/reconcile/openshift_base.py#L75-L80) as an error in the ResourceInventory, but reconciliation is progressing with other resources and clusters
* when all remaining work is done, the integration [checks for registered problems](https://github.com/geoberle/qontract-reconcile/blob/c4405e8d90ce45dd371f6f53cc06e8424d84e0af/reconcile/openshift_resources_base.py#L713-L714) in the ResourceInventory and fails if any have been found

for openshift-saas-deploy, the general path is the same

Reply:
I see, so it's really up to each oc_map caller to handle the case, often via the resourceinventory error handling. This allows to go as far as possible.
Makes sense, thanks!